user_tasks/signals.py

```diff
@@ -0,0 +1,174 @@
+
+from celery.signals import before_task_publish, task_failure, task_prerun, task_retry, task_success
+
+from .exceptions import TaskCanceledException
+from .models import UserTaskStatus
+from .tasks import UserTaskMixin
+
+LOGGER = logging.getLogger(__name__)
+
+
+@before_task_publish.connect
+def create_user_task(sender=None, body=None, **kwargs):  # pylint: disable=unused-argument
+    """
+    Create a :py:class:`UserTaskStatus` record for each :py:class:`UserTaskMixin`.
+
+    Also creates a :py:class:`UserTaskStatus` for each chain, chord, or group containing
+    the new :py:class:`UserTaskMixin`.
+    """
+    try:
+        task_class = import_string(sender)
+    except ImportError:
+        return
+    if issubclass(task_class.__class__, UserTaskMixin):
+        if 'user_id' in body['kwargs']:
+            user_id = body['kwargs']['user_id']
+        elif body['args']:
+            user_id = body['args'][0]
+        else:
+            raise TypeError('Each invocation of a UserTask must include the user_id')
+        try:
+
+            get_user_model().objects.get(pk=user_id)
+        except (ValueError, get_user_model().DoesNotExist):
+            raise TypeError('Invalid user_id: {}'.format(user_id))
+        task_id = body['id']
+        # Is it part of a chain?
+        if body.get('callbacks', []):
+            return create_chain_entry(user_id, task_id, task_class, body['args'], body['kwargs'], body['callbacks'])
+        if body.get('chord', None):
+            return create_chord_entry(task_id, task_class, body, user_id)
+        arguments_dict = task_class.arguments_as_dict(*body['args'], **body['kwargs'])
+        name = task_class.generate_name(arguments_dict)
+        total_steps = task_class.calculate_total_steps(arguments_dict)
+        parent = None
+        parent_class = ''
+        # Is it part of a group?
+        parent_id = body.get('taskset', None)
+        if parent_id:
+            parent_class = 'celery.group'
+        if parent_id:
+            parent_name = body['kwargs'].get('user_task_name', '')
+            parent, _ = UserTaskStatus.objects.get_or_create(
+                task_id=parent_id, defaults={'is_container': True, 'name': parent_name, 'task_class': parent_class,
+                                             'total_steps': 0, 'user_id': user_id})
+            if parent_name and not parent.name:
+                parent.name = parent_name
+                parent.save(update_fields={'name', 'modified'})
+        UserTaskStatus.objects.get_or_create(
+            task_id=task_id, defaults={'user_id': user_id, 'parent': parent, 'name': name, 'task_class': sender,
+                                       'total_steps': total_steps})
+        if parent:
+            parent.increment_total_steps(total_steps)
+
+
+def create_chain_entry(user_id, task_id, task_class, args, kwargs, callbacks, parent=None):
+    """Create and update status records for a new :py:class:`UserTaskMixin` in a Celery chain."""
```

Comment:
I originally followed that, but pydocstyle complained with `D200: One-line docstring should fit on one line with quotes`.  I'll switch them back and disable that warning.